Support calling torch.compile inside torch_dispatch

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #166417

When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    @classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @chenyang78

tugsbayasgalan committed:
Support calling torch.compile inside torch_dispatch

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

ezyang commented on torch/utils/_python_dispatch.py:
```diff
@@ -106,6 +124,23 @@ def _lazy_init_old_dispatch_mode_flags(self):
                 bool
             ] = deque()
 
+    def __init_subclass__(cls, **kwargs):
```
spooooooky

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -106,6 +124,23 @@ def _lazy_init_old_dispatch_mode_flags(self):
+            # _disable doesn't work on classmethods. This later fails in runtime anyways
+            if not isinstance(raw, classmethod):
+                wrapped = torch._disable_dynamo(raw, recursive=True)
+                cast(Any, cls).__torch_dispatch__ = wrapped
```
flexattention has always fallen back to eager under a torchdispatch mode, unless one sets compile_internals=True. This behavior is undesirable but it is the status quo

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -112,6 +130,23 @@ def _lazy_init_old_dispatch_mode_flags(self):
+                wrapped = torch._disable_dynamo(raw, recursive=True)
+                cast(Any, cls).__torch_dispatch__ = wrapped
+
+    @torch._disable_dynamo
```
The new behavior sounds like we want Dynamo to be disabled in all torch_dispatch. Can we just do this in Dynamo instead of complicate whatever is going on in TorchDispatchMode?

tugsbayasgalan replied:
I think we need a recursive disable. But does being recursive make any difference if it is in torch dispatch or dynamo?

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zou3519 reviewed (commented):
I don't think it is worth complicating the TorchDispatchMode stuff for this. Maybe we just take the hit and call this expected behavior then.

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -87,7 +104,14 @@ class TorchDispatchMode:
+            raw = cls.__dict__["__torch_dispatch__"]
+            if not isinstance(raw, classmethod):
+                cls.__torch_dispatch__ = torch._disable_dynamo(raw, recursive=True)
+
```
(or just an overall config to turn the whole thing on/off, but that means we need to keep the old check around)

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support calling torch.compile inside torch_dispatch"


When we are calling torch.compile inside torch_dispatch mode, we should enable it when there is no active custom dispatch mode that doesn't have ignore_compile_internal=True on the stack. See the test cases for various scenarios. For example:
```
backend3 = torch._dynamo.testing.EagerAndRecordGraphs()
backend2 = torch._dynamo.testing.EagerAndRecordGraphs()
backend = torch._dynamo.testing.EagerAndRecordGraphs()

class YoloMode2(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(
            lambda x, y: func(x, y), backend=backend3, fullgraph=True
        )(*args, **kwargs)
        return out

    classmethod
    def ignore_compile_internals(cls):
        return True

class YoloMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        out = torch.compile(torch.add, backend=backend2, fullgraph=True)(
            args[0], args[1]
        )
        return out

x = torch.ones(5)
with YoloMode(), YoloMode2():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)

self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 0)
self.assertEqual(len(backend.graphs), 0)
```

In the above example, we turn off outermost compile because there is at least one active dispatch mode that doesn't have ignore_compile_internals=True. We turn off torch.compile inside YoloMode2 because Yolomode is active and it doesn't ignore compile internals. We don't turn off torch.compile inside YoloMode because we don't have active mode anymore. 

Now let's say we switch the order and do following:
```
with YoloMode2(), YoloMode():
    torch.compile(
        lambda x, y: torch.add(x, y), fullgraph=True, backend=backend
    )(x, x)
```

Then we expect:
```
self.assertEqual(len(backend2.graphs), 1)
self.assertEqual(len(backend3.graphs), 1)
self.assertEqual(len(backend.graphs), 0)
```

outer-most torch.compile is ignored because there is at least one active mode. We compile YoloMode2 because even though YoloMode's active and it explicitly ignores compile internals. When we hit YoloMode, no mode is active so we compile. 

For context: https://github.com/pytorch/pytorch/issues/155331 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo chenyang78

[ghstack-poisoned]
